### PR TITLE
Add reshape op support for ui8.

### PR DESCRIPTION
**Pull request by ddilbazTT**
### Ticket
https://github.com/tenstorrent/tt-torch/issues/1168

### Problem description
Pytorch-xla converts boolean logical operations with ui8. 
For example, the following torch operation:
```
def test_booleanAdd():
    class Basic(nn.Module):
        def __init__(self):
            super().__init__()

        def forward(self, x, y):
            return x & y
    x = torch.tensor([True, False, True])
    y = torch.tensor([False, True, True])
    tt_mod = torch.compile(Basic(), backend="tt-experimental")
    out = tt_mod(x, y)
```

is expressed as:
```
module @SyncTensorsGraph.8 attributes {mhlo.cross_program_prefetches = [], mhlo.input_output_alias = [], mhlo.is_dynamic = false, mhlo.use_auto_spmd_partitioning = false} {
  func.func @main(%arg0: tensor<3xi1>, %arg1: tensor<3xi1>) -> tensor<3xi1> {
    %c = stablehlo.constant dense<0> : tensor<3xui8>
    %0 = stablehlo.convert %arg1 : (tensor<3xi1>) -> tensor<3xui8>
    %1 = stablehlo.convert %arg0 : (tensor<3xi1>) -> tensor<3xui8>
    %2 = stablehlo.and %0, %1 : tensor<3xui8>
    %3 = stablehlo.compare  NE, %2, %c : (tensor<3xui8>, tensor<3xui8>) -> tensor<3xi1>
    return %3 : tensor<3xi1>
  }
}
```

gpt-oss-20b model results in a boolean logic operation followed by a reshape, resulting in the following reshape operation:
```
%83 = stablehlo.reshape %82 : (tensor<71x1xui8>) -> tensor<71xui8>
```

Currently, reshape op does not support ui8. We need gpt-oss-20b support in the September 15 release, so we need a ui8 workaround for reshape op.

### What's changed
Added typecasts before and after reshape operation if the input is a ui8 through a TTNN workaround.

### Checklist
- [X] New/Existing tests provide coverage for changes

### CODE REVIEW

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp [hunk 1]
```diff
@@ -660,9 +660,50 @@ class ReshapeOpConversionPattern : public OpConversionPattern<ttir::ReshapeOp> {
   LogicalResult
   matchAndRewrite(ttir::ReshapeOp op, OpAdaptor adaptor,
                   ConversionPatternRewriter &rewriter) const override {
-    rewriter.replaceOpWithNewOp<ttnn::ReshapeOp>(
-        op, this->getTypeConverter()->convertType(op.getType()),
-        adaptor.getInput(), adaptor.getShape(), /*memory_config=*/nullptr);
+    auto inputType = mlir::cast<RankedTensorType>(adaptor.getInput().getType());
+    auto resultType = mlir::cast<RankedTensorType>(op.getType());
+
+    // Get input layout and data type
```
Comment (mmanzoorTT) [inline lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp]: Is there specific reason for adding typecasting here instead of TTNN Workarounds?

Comment (ddilbazTT) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp]: There isn't - let me know if I should make it a ttnn workaround

Comment (anuragsingh-tt) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp]: Yes, let's do that. I have a TT-Metal issue here: https://github.com/tenstorrent/tt-metal/issues/27843
Please link it in the reshape TTNN workaround.
You can use this as reference: https://github.com/tenstorrent/tt-mlir/pull/4399/files
If possible, can you also add a test in the builder? (test_ttir_ops.py)

Comment (mmanzoorTT) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp]: I agree, you should make it a ttnn workaround.

Comment (ddilbazTT) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp]: Done! I would appreciate it if you could review again.

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp [hunk 2]
```diff
@@ -663,6 +663,7 @@ class ReshapeOpConversionPattern : public OpConversionPattern<ttir::ReshapeOp> {
     rewriter.replaceOpWithNewOp<ttnn::ReshapeOp>(
         op, this->getTypeConverter()->convertType(op.getType()),
         adaptor.getInput(), adaptor.getShape(), /*memory_config=*/nullptr);
+
```
Comment (mmanzoorTT) [inline lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp]: Please revert this change.

File: lib/Dialect/TTNN/IR/TTNNWorkaroundsPass.cpp [hunk 3]
```diff
@@ -349,6 +349,25 @@ TTNNOperandsWorkaroundsFactory::createWhereOpOperandsWorkarounds(
       .addOutputOperandWorkaround(outputTypeWorkaround);
 }
 
+// Factory method to create a set of workarounds for reshape operation operands.
+// Reshape op only does not work with ui8 - force to bfloat16 then typecast
+// separately.
```
Comment (mmanzoorTT) [inline lib/Dialect/TTNN/IR/TTNNWorkaroundsPass.cpp:354]: Add relevant tt-metal issue here.

File: lib/Dialect/TTNN/IR/TTNNWorkaroundsPass.cpp [hunk 4]
```diff
@@ -349,6 +349,25 @@ TTNNOperandsWorkaroundsFactory::createWhereOpOperandsWorkarounds(
...
+TTNNOperandsWorkarounds
+TTNNOperandsWorkaroundsFactory::createReshapeOpOperandsWorkarounds(
+    RankedTensorType inputType) {
+  mlir::Type inputElementType = inputType.getElementType();
+  TTNNOperandWorkarounds typeWorkarounds;
+  mlir::tt::ttcore::DataType dataType =
+      mlir::tt::ttcore::elementTypeToDataType(inputElementType);
+  if (dataType == mlir::tt::ttcore::DataType::UInt8) {
+    typeWorkarounds.tensorDataTypeWorkaround =
+        mlir::tt::ttcore::DataType::BFloat16;
```
Comment (mmanzoorTT) [inline lib/Dialect/TTNN/IR/TTNNWorkaroundsPass.cpp]: I will suggest to typecast it to `int32` to avoid any loss between int->float and then float->int conversion. Secondly, the reshape op output is later used for `bitwise` op (as we have seen it in the stablehlo graph of the model) and `bitwise` op will again add a typecast to convert it to int32.

Comment (anuragsingh-tt) [reply lib/Dialect/TTNN/IR/TTNNWorkaroundsPass.cpp]: > I will suggest to typecast it to int32 to avoid any loss between int->float and then float->int conversion.

ui8 (0-255) -> bf16 is lossless?

Comment (ddilbazTT) [reply lib/Dialect/TTNN/IR/TTNNWorkaroundsPass.cpp]: I think Asif is right, but for full clarity isn't is because of ttnn bitwise ops? I don't think stablehlo bitwise ops are using i32.

Comment (mmanzoorTT) [reply lib/Dialect/TTNN/IR/TTNNWorkaroundsPass.cpp]: stablehlo bitwise ops uses any integer type but TTNN workaround typecast the inputs to int32 for bitwise ops. 
```
if (isa<ttnn::BitwiseAndOp, ttnn::BitwiseOrOp, ttnn::BitwiseXorOp>(op)) {
    if (dType == mlir::tt::ttcore::DataType::Int32) {
      return {};
    }
    return mlir::tt::ttcore::DataType::Int32;
  }
```

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp (hunk 1 above)
Comment (ddilbazTT) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp]: @anuragsingh-tt @mmanzoorTT I was wondering if you could help me out with the following error message:

```
ERROR: test=ttnn/test_reshape[ui8-shapes6]_ttnn.mlir.ttnn experienced an error with exception=Failed: program-level output golden comparison failed, actual_pcc=-0.0057108254054071324 < expected_pcc=0.99
```

Comment (ddilbazTT) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp]: Ignore my last comment please, that was incorrect

Comment (ddilbazTT) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp]: So it seems the workaround is applied:

```
#dram = #ttnn.buffer_type<dram>
#system_desc = #ttcore.system_desc<[{role = host, target_triple = "x86_64-pc-linux"}], [{arch = <wormhole_b0>, grid = 10x8, coord_translation_offsets = 18x18, l1_size = 1499136, num_dram_channels = 12, dram_channel_size = 1073741824, noc_l1_address_align_bytes = 16, pcie_address_align_bytes = 32, noc_dram_address_align_bytes = 32, l1_unreserved_base = 99712, erisc_l1_unreserved_base = 98304, dram_unreserved_base = 32, dram_unreserved_end = 1073077056, supported_data_types = [<f32>, <f16>, <bf16>, <bfp_f8>, <bfp_bf8>, <bfp_f4>, <bfp_bf4>, <bfp_f2>, <bfp_bf2>, <u32>, <u16>, <u8>, <si32>], supported_tile_sizes = [ 4x16,  16x16,  32x16,  4x32,  16x32,  32x32], dst_register_size_tiles = 8, num_cbs = 32, num_compute_threads = 1, num_datamovement_threads = 2}], [0], [1 : i32], [ 0x0x0x0]>
#ttnn_layout = #ttnn.ttnn_layout<(d0, d1) -> (d0, d1), <1x1>, memref<4x4x!ttcore.tile<32x32, u8>, #dram>, <interleaved>>
#ttnn_layout1 = #ttnn.ttnn_layout<(d0) -> (0, d0), <1x1>, memref<1x512x!ttcore.tile<32x32, u8>, #dram>, <interleaved>>
#ttnn_layout2 = #ttnn.ttnn_layout<(d0, d1) -> (d0, d1), <1x1>, memref<4x4x!ttcore.tile<32x32, si32>, #dram>, <interleaved>>
#ttnn_layout3 = #ttnn.ttnn_layout<(d0) -> (0, d0), <1x1>, memref<1x512x!ttcore.tile<32x32, si32>, #dram>, <interleaved>>
module {
  ttcore.device_module {
    builtin.module attributes {ttcore.system_desc = #system_desc} {
      ttcore.device @default_device = <workerGrid = #ttcore.grid<10x8, (d0, d1) -> (0, d0, d1)>, l1Map = (d0, d1, d2)[s0] -> (0, d0, d1, d2 + s0), dramMap = (d0, d1, d2)[s0, s1, s2, s3, s4, s5, s6] -> (0, 0, (((d0 * s1) * (s2 * (s3 * s6)) + d1 * (s2 * (s3 * s6)) + d2) floordiv s4) mod 12, ((((d0 * s1) * (s2 * (s3 * s6)) + d1 * (s2 * (s3 * s6)) + d2) floordiv s4) floordiv 12) * s4 + ((d0 * s1) * (s2 * (s3 * s6)) + d1 * (s2 * (s3 * s6)) + d2) mod s4 + s5), meshShape = 1x1, chipIds = [0]>
      func.func @reshape_wrapper(%arg0: tensor<128x128xui8, #ttnn_layout>) -> tensor<16384xui8, #ttnn_layout1> {
        %0 = "ttnn.typecast"(%arg0) <{dtype = #ttcore.supportedDataTypes<si32>}> : (tensor<128x128xui8, #ttnn_layout>) -> tensor<128x128xsi32, #ttnn_layout2>
        "ttnn.deallocate"(%arg0) <{force = false}> : (tensor<128x128xui8, #ttnn_layout>) -> ()
        %1 = "ttnn.reshape"(%0) <{shape = [16384 : i32]}> : (tensor<128x128xsi32, #ttnn_layout2>) -> tensor<16384xsi32, #ttnn_layout3>
        "ttnn.deallocate"(%0) <{force = false}> : (tensor<128x128xsi32, #ttnn_layout2>) -> ()
        %2 = "ttnn.typecast"(%1) <{dtype = #ttcore.supportedDataTypes<u8>}> : (tensor<16384xsi32, #ttnn_layout3>) -> tensor<16384xui8, #ttnn_layout1>
        "ttnn.deallocate"(%1) <{force = false}> : (tensor<16384xsi32, #ttnn_layout3>) -> ()
        return %2 : tensor<16384xui8, #ttnn_layout1>
      }
    }
  }
}
``` 
This is: `ttnn/test_reshape[ui8-shapes0]_ttnn.mlir`

Asif later debugged that this is a ttrt level problem. The inputs for ttrt seem to be float32. I will create an issue and mark `pytest.skip` for now.

Comment (ddilbazTT) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp]: https://github.com/tenstorrent/tt-mlir/issues/4813 - Created the issue.

File: test/ttmlir/Silicon/TTNN/n150/data_movement/reshape/reshape.mlir [hunk 5]
```diff
@@ -22,3 +22,12 @@ func.func @reshape_to_scalar(%arg0: tensor<1xf32>) -> tensor<f32> {
   // CHECK: "ttnn.reshape"
   return %1 : tensor<f32>
 }
+
+func.func @reshape_ui8(%arg0: tensor<71x1xui8>) -> tensor<71xui8> {
```
Comment (sdjordjevicTT) [inline test/ttmlir/Silicon/TTNN/n150/data_movement/reshape/reshape.mlir]: Can you just extract and move this test under:
test/ttmlir/Dialect/TTNN/Transforms/Workarounds

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp (hunk 1 above)
Comment (mmanzoorTT) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp]: @tapspatel has suggested a solution. Please have a look and fix the builder test.

File: lib/Dialect/TTNN/IR/TTNNWorkaroundsPass.cpp [hunk 6]
```diff
@@ -349,6 +349,26 @@ TTNNOperandsWorkaroundsFactory::createWhereOpOperandsWorkarounds(
       .addOutputOperandWorkaround(outputTypeWorkaround);
 }
 
+// Factory method to create a set of workarounds for reshape operation operands.
+// Reshape op only does not work with ui8 - force to bfloat16 then typecast
```
Comment (mtopalovicTT) [inline lib/Dialect/TTNN/IR/TTNNWorkaroundsPass.cpp]: nit: Force to int32?

**Review by mtopalovicTT — APPROVED** [review]

**Review by anuragsingh-tt — APPROVED** [review]

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp (hunk 1 above)
Comment (ddilbazTT) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp]: @mmanzoorTT I tried but it seems like there was a misunderstanding. His solution doesn't apply in this case. For context, this is what I tried:

```
def reshape(
    in0: Operand,
    builder: TTIRBuilder,
    input_shape: Shape,
    output_shape: Shape,
    dtype: torch.dtype,
):
    input_tensor = torch.ones(input_shape, dtype=dtype)
    
    golden_output_tensor = input_tensor.reshape(output_shape).to(dtype)
    
    builder.set_graph_input_output(
        [input_tensor], [golden_output_tensor], override=True
    )
    return builder.reshape(in0, output_shape)


@pytest.mark.parametrize(
    "shapes",
    [
        # [input_shape, output_shape]
        [(128, 128), (16384,)],  # Flatten 2D to 1D
        [(24,), (2, 3, 4)],  # Unflatten 1D to 3D
        [(2, 3, 4), (6, 4)],  # 3D to 2D reshape
        [(128, 128), (64, 256)],  # 2D to 2D different arrangement
        [(1, 1, 1), (1,)],  # Edge case: all dimensions are 1
        [(10,), (10,)],  # Identity reshape
        [(64, 512), (64, 1, 512)],  # Common ML pattern: expand dims
        [(256, 256), (512, 128)],  # Power of 2 reshape
        [(32, 3, 224, 224), (32, 150528)],  # Large ML pattern: batch flatten
    ],
)
@pytest.mark.parametrize(
    "dtype", [torch.float32, torch.int32, torch.uint8], ids=["f32", "i32", "ui8"]
)
def test_reshape(shapes, dtype: torch.dtype, request):
    input_shape, output_shape = shapes

    def reshape_wrapper(in0: Operand, builder: TTIRBuilder):
        return reshape(in0, builder, input_shape, output_shape, dtype)

    compile_ttir_to_flatbuffer(
        reshape_wrapper,
        [input_shape],
        [dtype],
        test_base=request.node.name,
        output_root=request.config.getoption("--path"),
        system_desc_path=request.config.getoption("--sys-desc"),
    )
```

They need to fix this in ttrt because from the builder, we can't pass in any other data type for the golden values.

Comment (tapspatel) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp]: Sorry for the confusion, we don't support golden types other than float32 in builder. It is a severe limitation that we will address in that issue. You can still set golden inputs/outputs otherwise (as long as it's float32).
